Add a contributor guide

ehuss opened the pull request:
This adds a new contributor guide. It is built with mdbook with the intention that it is published on GitHub Pages. This replaces several of the loosely organized documents that we have accumulated over time. It also adds some new content aimed at contributors, such as an overview of the contribution process. 

Several of the chapters are essentially stubs or basic starting points with the intention that they get reworked in the future. There's also still a lot of work left to resolve the [Meta issues](https://github.com/rust-lang/reference/issues?q=is%3Aissue%20state%3Aopen%20label%3AMeta).

Note that I am also working on a separate but closely related initiative to add a scope chapter to the Reference itself, along with a rework of the introduction. This is intended to further clarify to contributors (and readers!) what is expected of the content.

ehuss committed:
Start a contributor guide

This is the beginning of a new mdbook book that will house all of the
guidelines for contributors. This is published via GitHub Pages.

ehuss committed:
Move the building section to the contributor guide

ehuss committed:
Apply some light edits to "Building the Reference"

ehuss committed:
Move the publishing process to the contributor guide

ehuss committed:
Move the contributing guide to the introduction

ehuss committed:
Add a forward to the new contributing guide

ehuss committed:
Unwrap the introduction

ehuss committed:
Apply some light edits to the introduction

This is just some light editing. I expect that this chapter will have
larger edits in the future, but I want to defer that till later.

ehuss committed:
Move the review policy to the contributor guide

ehuss committed:
Apply some light edits to the review policy

ehuss committed:
Move the grammar guide to the contributor guide

ehuss committed:
Apply some light edits to the rust grammar chapter

ehuss committed:
Move the attribute template to the contributor guide

ehuss committed:
Update attributes chapter title

ehuss committed:
Add a chapter describing mdbook-spec

ehuss committed:
Start a formatting chapter

ehuss committed:
Move markdown formatting to the contributor guide

ehuss committed:
Move admonitions to the contributor guide

ehuss committed:
Move examples to the contributor guide

ehuss committed:
Move standard library links to the links chapter

ehuss committed:
Add more details to the links chapter

ehuss committed:
Move rules to the contributor guide

ehuss committed:
Add a contributor tests chapter

ehuss committed:
Start a style guide

ehuss committed:
Start the wording style chapter

This is just a stub, with the expectation that it will be
expanded/rewritten later.

ehuss committed:
Start the punctuation chapter

This is just a stub, with the expectation that it will be
expanded/rewritten later.

ehuss committed:
Add an introduction to the tooling sub-chapters

ehuss committed:
Add a resources chapter

ehuss committed:
Link README to the new contributor guide

ehuss committed:
Remove STYLE.md

This has been superseded by the contributor guide.

ehuss committed:
Add a contribution process chapter

ehuss committed:
Apply some light edits to the rules chapter

ehuss committed:
Apply some light edits to the publishing process

ehuss committed:
Apply light edits to standard library links

ehuss committed:
Apply some light edits to the introduction

This introduction should probably receive deeper edits to make it fit
into the guide better.

ehuss committed:
Apply some light edits to the grammar chapter

ehuss committed:
Apply some light edits to the markdown chapter

ehuss committed:
Apply some light edits to the examples chapter

ehuss committed:
Apply light edits to the attributes chapter

tshepang commented on .github/workflows/main.yml:
```diff
@@ -107,6 +107,23 @@ jobs:
       run: cargo fmt --check
 
+  contributor-guide:
+    name: Contributor-guide build check
```
```suggestion
    name: Contributor Guide build check
```

traviscross committed:
Fix restrictive "which" to "that"

These are restrictive clauses that identify specific things
(the extension that adds features, the admonitions that use a style,
the appendix that tracks rules), so they take "that" rather than
"which".

traviscross committed:
Fix "like" to "such as" for examples

When introducing examples, "such as" is correct; "like" implies
resemblance rather than membership.  Also fixes "em-dash" to "em
dash" (two words, no hyphen) per Chicago Manual of Style.

traviscross committed:
Fix grammar in rules chapter

Remove commas from compound predicates (two verbs sharing one subject
don't take a comma before the conjunction).  Add missing comma after
"e.g." per standard punctuation for Latin abbreviations.  Also change
"like" to "e.g.," in the rule name example, since this introduces an
example.

traviscross committed:
Normalize sublist indentation

The SUMMARY.md was already changed to use 2-space indentation for
sublists.  This brings the remaining files into consistency with that
direction, which is to use the minimum indentation required by
CommonMark.

traviscross committed:
Fix review policy style issues

Change asterisk list markers to dashes for consistency with the rest
of the guide.  Replace "Language" with "Prose" in the
Understandability bullet to avoid a miscue (this is a guide about a
programming language, so using "language" to mean "text" is
confusing).  Replace "content in the spec" with "content in the
Reference" for consistency.  Rephrase "considers themselves
well-versed enough in the topic to understand it" to avoid the
singular-they issue and tighten the wording.  Remove trailing space
from a heading.

traviscross committed:
Fix introduction.md style issues

Replace the awkward passive "It is encouraged for you to read" with
the direct "We encourage you to read".  Remove the unused link
definition for Authoring Guide, which is never referenced in the text.

traviscross committed:
Use bold title-case link for Review Policy

Per our style direction, when linking to a chapter by name, use title
case with bold for external docs.

traviscross committed:
Capitalize Reference in publishing.md

"The Reference" is a proper noun throughout the guide and should be
capitalized consistently.

traviscross committed:
Use bold title-case link for Building chapter

Replace "the building chapter" with the bold title-case form "Building
the Reference" per our style direction for chapter links.  This also
avoids the redundant "the ...  chapter" phrasing.

traviscross committed:
Fix codeblocks to code blocks

"Code blocks" is two words.

traviscross committed:
Fix admonitions.md grammar and link

The compound subject "color and styling" takes a plural verb: "are
defined" rather than "is defined".  Also change blob/master to
blob/HEAD to match the other link on the same line and the convention
used elsewhere in the guide.

traviscross committed:
Fix double space after period in links.md

Only one space after a sentence is accepted.

traviscross committed:
Fix malformed table rows in grammar.md

Two table rows were missing their closing pipe delimiter, which made
the rows malformed.

traviscross committed:
Clarify ambiguous pronoun in attributes.md

Replace "they" with "the examples" to resolve the ambiguity over
whether the pronoun refers to the examples or the phrasing.

traviscross committed:
Tighten wording in resources.md

Replace "very useful to become familiar with" with the more direct
"useful to learn".

traviscross committed:
Fix vague pronoun in punctuation.md

Replace "rewrite it" with "rewrite the phrase" to clarify what should
be rewritten.  Also fix the plural "conjunctions" to the singular "a
conjunction" since the advice applies to a single slash.

traviscross committed:
Disambiguate sentence in stabilization.md

The original "a section of the Reference or language" could be read as
"a section of (the Reference or language)" or as "(a section of the
Reference) or (language)".  Using em dashes makes the parenthetical
"or a part of the language" unambiguous.

traviscross committed:
Fix inconsistent em dash hyphenation

The first occurrence on this line read 'em-dash' (hyphenated) while
the second read 'em dash' (two words).  Per the Chicago Manual of
Style, 'em dash' is the standard form.  Let's use it consistently.

traviscross committed:
Normalize sublist indentation to 2 spaces

mdbook tolerates both 2-space and 4-space indentation for sublists,
but 2 spaces is the more common convention and is what we use
elsewhere.  Let's normalize all sublist items in SUMMARY.md and in the
contribution process overview to use 2-space indentation consistently.

traviscross committed:
Apply light edits to the README

The opening paragraph of the README has some filler words and a run-on
sentence.  This tightens the prose by removing the unnecessary
appositive ("using the [mdbook] tool to convert to HTML") and
splitting one long sentence into two shorter ones joined by a
semicolon.

traviscross committed:
Apply light edits to the introduction

This makes several editorial improvements to the introduction:

- Bold the title ("the Rust Reference!" becomes "**The Rust
  Reference**.") and change "guide" to "document".  Remove
  the comma before the compound predicate "and serves".

- Tighten the glossary sentence to be less verbose.

- Tighten "the hardest task, but definitely the most valuable"
  to "the hardest but most valuable task".  Remove the comma
  before the compound predicate "and write about it".

- Fix the branch name from `master` to `main`, matching the
  current name of the primary branch in rust-lang/rust.

- Use the bold-link convention for The Unstable Book, matching
  the style used for other important cross-references.

traviscross committed:
Apply light edits to the contribution process

This makes several editorial improvements to the contribution
process chapter:

- Qualify the advice to discuss changes before opening a PR by
  adding "for nontrivial changes" so as not to discourage
  small drive-by fixes.

- Use bold title-case links for Publishing, Issue Assignment,
  and Stabilization, matching the cross-reference convention
  used elsewhere in the guide.

- Use em dash parenthetical lists ("--- such as ... ---")
  instead of bare commas for the minor-changes and
  major-changes descriptions.

- Tighten prose throughout: "for more direct discussions"
  becomes "for discussions"; "Try to keep" becomes "Keep";
  fix pronoun agreement for "it passes" to "these pass".

- Replace "like" with "such as" when introducing examples,
  which is the more precise usage.

traviscross committed:
Fix conflation in stabilization exception sentence

The sentence describing the exception to the stabilization process
conflates two different things.  It says 'a section of the Reference
--- or a part of the language --- that is undocumented', applying
'undocumented' to both.  But only the language can be undocumented; a
section of the Reference, by contrast, can be incomplete.

Let's separate these into their correct forms: 'a part of the language
that is undocumented or a section of the Reference that is
incomplete'.

traviscross committed:
Refine the undocumented-feature exception

The exception paragraph says that changes to undocumented or
incomplete sections don't need a Reference update, but this is too
broad.  It's true for changes to the details of something that isn't
documented, but when a new feature introduces a rule that can be
stated independently, that rule should still be added to the relevant
placeholder section.

Let's split this into two paragraphs: the first keeps the existing
exception for changes to undocumented details, and the second
clarifies that independently stateable rules should be documented even
in placeholder sections.

traviscross committed:
Fix "pseudo-code" to "pseudocode"

The prefix "pseudo-" generally joins to the base word without a
hyphen.  "Pseudocode" is one word in standard usage.

traviscross committed:
Fix "re-exports" to "reexports"

Words formed with the prefix "re-" generally take no hyphen.

traviscross committed:
Remove unnecessary intensifier "very"

The word "very" adds no meaning here.  "Welcome" already conveys the
intended sentiment.

traviscross committed:
Capitalize "Markdown" as a proper noun

Markdown is a proper noun (the markup language created by John
Gruber).  This fixes six instances in prose where "markdown" appears
in lowercase: one in tests.md, four in grammar.md, and one in
admonitions.md.  We leave lowercase instances inside code-block
language tags and file extensions unchanged.

traviscross committed:
Rewrite tooling chapter for conciseness

The tooling index page had several verbose constructions:
"Additionally, there is... that adds a variety of", "In addition to
the book tooling, there is a variety of...".  This rewrites the page
for fluency and replaces "a variety of" with "several" in both
tooling/index.md and tooling/building.md.

We capitalize the cross-references to Building the Reference and
Running Tests as titles but do not bold them, since both targets are
within the dev-guide (internal links).  We also rename the "tests
chapter" link label to "Running Tests" to match the target page title,
following the convention of not saying "the X chapter".

traviscross committed:
Replace "for its usage" with "for more information"

Our upcoming style guide reserves "usage" for customary patterns or
measurable consumption.  Here the intent is simply to point the reader
to more details, so "for more information" is both more natural and
more accurate.

traviscross committed:
Fix chapter cross-reference in Running Tests

Our convention is to name the chapter by its title rather than saying
"the X chapter".  We capitalize the link text as a title but do not
bold it because the target (examples.md) is within the dev-guide (an
internal link).

traviscross committed:
Fix section cross-reference in Admonitions

Our convention is to name the section by its title rather than saying
"the X section".  We capitalize the link text as a title and bold it
because the target is in the Reference proper (an external book).  We
also add a brief locator ("in the Reference introduction") for
context, since the section name alone does not clearly indicate where
the content lives.

traviscross committed:
Fix chapter cross-reference in Rust Grammar

Our convention is to name the chapter by its title rather than saying
"the X chapter".  We capitalize the link text as a title and bold it
because the target is the Notation chapter of the Reference proper (an
external book).  Let's also update the link definition label to match.

traviscross committed:
Fix chapter cross-references in Markdown

Our convention is to name chapters by their titles rather than saying
"the X chapter".  We capitalize the link text as titles but do not
bold it, since both targets are within the dev-guide (internal links).
Let's also update the link definition labels to match.

traviscross committed:
Use precise CommonMark terms for link constructs

The CommonMark spec (§6.6) calls these "reference links" and "shortcut
reference links".  The original phrasing ("The use of reference links
is preferred, with shortcuts if appropriate") is slightly stilted and
uses the informal abbreviation "shortcuts".  This rewrites the
sentence for directness and uses the full CommonMark term "shortcut
reference links".

traviscross committed:
Clarify what need not be duplicated

The original sentence ("Some of this information may already
exist... so there is no need to duplicate it") is vague about what
"it" refers to and where duplication should be avoided.  Let's
clarify.

traviscross committed:
Clarify why inline tests fail for stabilization PRs

The original text says inline tests "will fail" without explaining why
or when.  The rewrite explains the mechanism.

traviscross committed:
Clarify rule name specificity in test annotations

The original phrasing ("Prefixed rule names should not be used in
tests") is unclear about what it means.  What it really means is that
tests should use the most specific rule name available rather than a
parent name.  The rewrite states this positively and gives a concrete
example.

traviscross committed:
Improve tone of test coverage expectations

The original phrasing ("Although it would be nice, it is unrealistic")
is a bit casual.  Let's sharpen it up.

traviscross committed:
Reword grammar self-description parenthetical

The original parenthetical ("written in itself, hopefully that's not
too confusing") is too casual and self-deprecating for a contributor
guide.  The rewrite conveys the same information -- that the grammar
notation describes itself using its own notation -- more clearly and
concisely.

traviscross committed:
Reword grammar prefix guidance for clarity

The original phrasing ("if you prefer to be more explicit") frames the
choice as a matter of personal preference.  The rewrite frames it as a
matter of serving the reader: use the prefix when explicitness would
aid clarity.  This better captures the intent.

traviscross committed:
Clarify expert consultation guidance

The original sentence has an ambiguous "or" that conflates two
conditions: "if the reviewer... is well-versed in the topic, or if the
relevant experts are the author or have been heavily involved in the
process."  The reader cannot tell whether "well-versed in the topic"
is an independent condition or part of the first clause.  The rewrite
separates the three conditions into a clear serial list and tightens
the phrasing.

traviscross committed:
Clarify minor content changes policy

The original text says minor changes "can be made with the review from
a team member without further consultation" but does not make clear
that a maintainer may push fixes directly to the PR branch and merge
without consulting the author.  The rewrite makes this workflow
explicit.

traviscross committed:
Remove outdated team member reference

pnkfelix is no longer on the spec team or the lang team.

traviscross committed:
Consolidate style guide into a single chapter

The style guide currently lives across three files (index.md,
wording.md, punctuation.md) plus an empty "Language and grammar"
placeholder.  The content is small enough that a single file is better
for now.

traviscross committed:
Bump to mdBook v0.5.2

There's a new mdBook v0.5.2 release.  It has various fixes we want.
Let's bump our CI to use it.
